docs/design/proposal-troubleshoot-consolidated-cli.md

```diff
@@ -0,0 +1,292 @@
+
+This new package should be kept as minimal as possible. and serve only as an interface to private functions in the other packages.
+
+Once the stable API is ready we can instruct projects like kurl to target that and work on removing code marked for deprecation and migrate non-public functions to an `internal` package to make it clear that it's not intended to be imported.
+
+The core of this api will be exposed via new troubleshoot `types`
+
+```go
+type TroubleshootKinds struct {
+}
+
+type Bundle struct {
+  FilePath string
+}
+
+type AnalysisResults struct {
+}
+
+```
+
+from there we can build out methods for each to interact with them:
+
+```go
+// Load loads specs defined by the LoadOptions struct and returns a TroubleshootKinds struct
+func Load(LoadOptions) (TroubleshootKinds,err){}
+kinds, err := Load(LoadOptions)
+
+// Bundle.Collect runs collections defined in TroubleshootKinds passed through CollectOptions
+func (bundle *Bundle) Collect(CollectOptions) (error) {}
+
+func (bundle *Bundle) Load(LoadBundleOptions) (error) {}
+
+// Bundle.Analyze runs analysis defined in TroubleshootKinds passed through AnalyzeOptions
+func (bundle *Bundle) Analyze(AnalyzeOptions) (AnalysisResults, error) {}
+
+// Bundle.Redact runs redaction defined in TroubleshootKinds passed through RedactOptions
+func (bundle *Bundle) Redact(RedactOptions) error {}
+
+// Bundle.Archive produces an archive from a bundle on disk with options passed in ArchiveOptions
+func (bundle *Bundle) Archive(ArchiveOptions) error {}
+
+// Bundle.Serve starts an sbctl-like server with options defined in ServeOptions
+func (bundle *Bundle) Serve(ServeOptions) error {}
+```
+
+To Allow for forward flexibility while reducing the impact of potentially breaking changes, options will be passed into these methods via options structs.
+
+```go
+type LoadOptions struct {
+  RawSpecs []string // list of locations for specs to load
+  RawSpec string // a single spec to parse
+  FilePaths []string // list of filepaths to check, can be globbed
+  URIs []string // list of URIs to retrieve specs from
+  SearchCluster bool // toggle for searching cluster from context for troubleshoot objects
+}
+
+type LoadBundleOptions struct {
+    Path string // Path to archive or directory of bundle files
+}
+
+type CollectOptions struct {
+  Specs TroubleshootKinds // list of specs to extract collectors and redactors from
+}
+
+type RedactOptions struct {
+  Specs TroubleshootKinds // list of specs to extract redactors from
+}
+
+// Note: this is almost identical to `CollectOptions` for now but remains separate to enable easier addition of redact specific options at a later date
+
+type ServeOptions struct {
+  Address string // address to listen on including port (0.0.0.0:8080)
+  ConfigPath string // optional path to store generated kubeconfig
+}
+```
+
```

Comment:
would love to see an example of how a bundle is created with a spec - I'm guessing something like

```go
mybundle, err = Bundle.Collect(kinds)
```